x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java

```diff
@@ -0,0 +1,389 @@
+        if (this.isLegacy != version.isLegacy) {
+            return this.isLegacy ? true : false;
+        }
+        return version.id >= id;
+    }
+
+    public boolean between(MlConfigVersion lowerInclusive, MlConfigVersion upperExclusive) {
+        if (upperExclusive.onOrBefore(lowerInclusive)) throw new IllegalArgumentException();
+        return onOrAfter(lowerInclusive) && before(upperExclusive);
+    }
+
+    public static MlConfigVersion fromVersion(Version version) {
+        return fromId(version.id);
+    }
+
+    public static Version toVersion(MlConfigVersion mlConfigVersion) {
+        if (mlConfigVersion.id < MINIMUM_COMPATIBLE.id) {
+            throw new IllegalArgumentException("Cannot convert " + mlConfigVersion.id + ". Incompatible version");
+        }
+        return Version.fromId(mlConfigVersion.id);
+    }
+
+    public static MlConfigVersion getMinMlConfigVersion(DiscoveryNodes nodes) {
+        return getMinMaxMlConfigVersion(nodes).v1();
+    }
+
+    public static MlConfigVersion getMaxMlConfigVersion(DiscoveryNodes nodes) {
+        return getMinMaxMlConfigVersion(nodes).v2();
+    }
+
+    public static Tuple<MlConfigVersion, MlConfigVersion> getMinMaxMlConfigVersion(DiscoveryNodes nodes) {
+        MlConfigVersion minMlConfigVersion = MlConfigVersion.CURRENT;
+        MlConfigVersion maxMlConfigVersion = MlConfigVersion.MINIMUM_COMPATIBLE;
+        for (DiscoveryNode node : nodes) {
+            MlConfigVersion mlConfigVersion = getMlConfigVersionForNode(node);
+            if (mlConfigVersion.before(minMlConfigVersion)) {
+                minMlConfigVersion = mlConfigVersion;
+            }
+            if (mlConfigVersion.after(maxMlConfigVersion)) {
+                maxMlConfigVersion = mlConfigVersion;
+            }
+        }
+        return new Tuple<>(minMlConfigVersion, maxMlConfigVersion);
+    }
+
+    public static MlConfigVersion getMlConfigVersionForNode(DiscoveryNode node) {
+        String mlConfigVerStr = node.getAttributes().get(ML_CONFIG_VERSION_NODE_ATTR);
+        if (mlConfigVerStr == null) {
+            return fromVersion(node.getVersion());
+        }
+        return fromString(mlConfigVerStr);
+    }
+
+    public static MlConfigVersion fromString(String str) {
+        if (str == null) {
+            return CURRENT;
+        }
+        if (str.contains(".")) {
+            return fromVersion(Version.fromString(str));
+        }
+        int versionNum = Integer.parseInt(str);
+        if (versionNum < V_10.id) {
+            throw new IllegalArgumentException("Illegal version " + versionNum);
+        }
+        return fromId(Integer.parseInt(str));
+    }
+
+    @Override
+    public int compareTo(MlConfigVersion other) {
+        if (this.isLegacy != other.isLegacy) {
+            return this.isLegacy ? -1 : 1;
+        }
+        return Integer.compare(this.id, other.id);
+    }
+
+    @Override
+    public String toString() {
+        if (id >= MINIMUM_COMPATIBLE.id) {
```

Comment:
I found this strange. It's a bit weird because we're using lower numbers for newer versions.

At the very least this deserves a comment here.

But I also wonder if the lack of comparability of the numeric IDs is going to bite us at some point. I think it would be better to use numeric IDs that are 10000000 greater than the printed number we'll use. Then the numeric IDs would compare naturally. In other words print 10000010 as "10".

Reply:
We've taken the approach of just keeping the same number going - so our detached versions start at 8_500_000 and go up from there. Then it's just comparing a single number

Reply:
For usage in config documents we have the additional requirement to be able to serialize to and from text.

This is why I was thinking that it would be nice to use smaller numbers for the printed version. So we'd progress 8.9.0, 8.9.1, 8.9.2, 10, 11, 12, 13, .... But then it would also be nice if the internal single number ID was incrementing too.

However, just thinking about this a bit more, we'll still need to print in x.y.z format to support rolling upgrades.  If we print "10" into a config document and an 8.9 node reads that then it will throw an exception: https://github.com/elastic/elasticsearch/blob/568b292bdeb04103cb7d30f191a6378bac8f57ed/server/src/main/java/org/elasticsearch/Version.java#L255-L259

That would happen somewhere like here: https://github.com/elastic/elasticsearch/blob/4c8847bdd4c31aa2ddb04b0d5eb7c2968696a316/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java#L874

So I am actually thinking the best way to print these numbers is with .0.0 appended, so progressing 8.9.0, 8.9.1, 8.9.2, 10.0.0, 11.0.0, 12.0.0, 13.0.0, .... The numbers starting 10 or above wouldn't be semantic versions, just incremented when necessary. But 8.9 or earlier nodes could parse them and would understand that they were greater than that node's version.